### ISSUE TITLE
ServiceManager::get results in endless loop since 3.11.0

### ISSUE THREAD

**Issue by InvisibleSmiley** [laminas/laminas-cache#284]
### BC Break Report

|    Q        |   A
|------------ | ------
| Version     | 3.11.0, 3.12.0
| Last known good     | 3.10.1

#### Summary

`ServiceManager::get` calls result in an endless loop unless the "config" service has been configured

#### Previous behavior

Works as expected

#### Current behavior

Endless loop

#### How to reproduce

```php
<?php

declare(strict_types=1);

use Laminas\Cache\ConfigProvider;
use Laminas\Cache\Service\StorageAdapterFactoryInterface;
use Laminas\ConfigAggregator\ConfigAggregator;
use Laminas\ServiceManager\ServiceManager;

require_once __DIR__ . '/vendor/autoload.php';

$configAggregator = new ConfigAggregator([ConfigProvider::class]);
$config = $configAggregator->getMergedConfig();
$container = new ServiceManager($config['dependencies']);

// Without this, the get() call below results in an endless loop with laminas/laminas-cache >= 3.11.0
//$container->setService('config', $config);

$container->get(StorageAdapterFactoryInterface::class);
```

**Comment by Ocramius** [comment]
Could this be put in a test case, to prevent any further regressions?

**Comment by InvisibleSmiley** [comment]
Limited stack trace (recursion happens quite early as you can see):
```
PHP Fatal error:  Uncaught Error: Xdebug has detected a possible infinite loop, and aborted your script with a stack depth of '512' frames in /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php:950
Stack trace:
(...)
#493 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(71): Laminas\ServiceManager\ServiceManager->has('config')
#494 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(35): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->getConfig(Object(Laminas\ServiceManager\ServiceManager))
#495 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(970): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->canCreate(Object(Laminas\ServiceManager\ServiceManager), 'config')
#496 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(297): Laminas\ServiceManager\ServiceManager->abstractFactoryCanCreate('config')
#497 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(71): Laminas\ServiceManager\ServiceManager->has('config')
#498 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(35): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->getConfig(Object(Laminas\ServiceManager\ServiceManager))
#499 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(970): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->canCreate(Object(Laminas\ServiceManager\ServiceManager), 'config')
#500 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(297): Laminas\ServiceManager\ServiceManager->abstractFactoryCanCreate('config')
#501 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(71): Laminas\ServiceManager\ServiceManager->has('config')
#502 /testproject/vendor/laminas/laminas-cache/src/Service/StorageCacheAbstractServiceFactory.php(35): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->getConfig(Object(Laminas\ServiceManager\ServiceManager))
#503 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(970): Laminas\Cache\Service\StorageCacheAbstractServiceFactory->canCreate(Object(Laminas\ServiceManager\ServiceManager), 'config')
#504 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(297): Laminas\ServiceManager\ServiceManager->abstractFactoryCanCreate('config')
#505 /testproject/vendor/laminas/laminas-cache/src/Service/StorageAdapterPluginManagerFactory.php(24): Laminas\ServiceManager\ServiceManager->has('config')
#506 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(642): Laminas\Cache\Service\StorageAdapterPluginManagerFactory->__invoke(Object(Laminas\ServiceManager\ServiceManager), 'Laminas\\Cache\\S...', NULL)
#507 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(236): Laminas\ServiceManager\ServiceManager->doCreate('Laminas\\Cache\\S...')
#508 /testproject/vendor/laminas/laminas-cache/src/Service/StorageAdapterFactoryFactory.php(15): Laminas\ServiceManager\ServiceManager->get('Laminas\\Cache\\S...')
#509 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(642): Laminas\Cache\Service\StorageAdapterFactoryFactory->__invoke(Object(Laminas\ServiceManager\ServiceManager), 'Laminas\\Cache\\S...', NULL)
#510 /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(236): Laminas\ServiceManager\ServiceManager->doCreate('Laminas\\Cache\\S...')
#511 /testproject/test.php(19): Laminas\ServiceManager\ServiceManager->get('Laminas\\Cache\\S...')
#512 {main}
  thrown in /testproject/vendor/laminas/laminas-servicemanager/src/ServiceManager.php on line 950

```

**Comment by boesing** [comment]
Hm, so "yes" this is most probably related to the feature added in 3.11 which adds the possibility to register storage adapters via config, but when I see that trace, I wonder if this is more related to the servicemanager.
IMHO it should handle recursions for both `has` and `get` to prevent this kind of problems.

This is mostly related to the abstract factory also checking for the `config` service (which is used to check `config` exists). 
So we can add a safe-guard check in the `StorageCacheAbstractServiceFactory` to verify that the service name called is not `config` but IMHO that would be too mezzio/laminas application specific and thus I am not sure if it has to be handled here.

WDYT @Ocramius

**Comment by Ocramius** [comment]
> but IMHO that would be too mezzio/laminas application specific and thus I am not sure if it has to be handled here.

that's kinda why I'd like to represent it here with a small test scenario: I also don't have the full picture, but a test makes this manageable / can fit my brain then :D

**Comment by Xerkus** [comment]
Any calls to container `has()` or `get()` for non-existing service from abstract factory `canCreate()` will result in endless loop. Actually it will loop if service exists but created by another abstract factory that is supposed to be asked next.
I don't think service manager can reasonably track it without massive overhead.

When `has()` call is used with fixed id it can be solved by factory itself by returning false when `canCreate()` is asked for that fixed id.

**Comment by boesing** [comment]
I don't think it is fixable in this project.
It might be fixable by adding a check to `has` in `ServiceManager` which tracks the service name for `has` for `abstract_factories` and skips those which are still used for determining if a service exists.

I have an idea in mind, I'll see if I can create a PoC. If someone has another idea, just let me know.

@InvisibleSmiley can you create a PR with just a failing test where the problem occurs? Just pass the code from the issue description to that test, I will handle service manager then. I already got my head wrapped around the issue but that would provide others to get into it.

**Comment by Xerkus** [comment]
It is fixable, sorta:

```diff
     public function canCreate(ContainerInterface $container, $requestedName)
     {
+        if ($requestedName === 'config') {
+            return false; // avoid endless loop when config is not registered in container and this factory is registered as abstract factory
+        }
         $config = $this->getConfig($container);
         if (empty($config)) {
             return false;
         }
         return isset($config[$requestedName]) && is_array($config[$requestedName]);
     }
```

**Comment by boesing** [comment]
That would fix the symptom for sure.
Do we want to apply this and fix the underlying issue in `laminas-servicemanager` afterwards?
Any vetos?

**Comment by Ocramius** [comment]
Meanwhile: #286

**Comment by boesing** [comment]
Closed with #286
